diff --git a/work/share/openPMD/cmake/FindADIOS.cmake.orig b/work/share/openPMD/cmake/FindADIOS.cmake
index 62b0b07..d3dfa87 100644
--- a/work/share/openPMD/cmake/FindADIOS.cmake.orig
+++ b/work/share/openPMD/cmake/FindADIOS.cmake
@@ -209,6 +209,7 @@ if(ADIOS_FOUND)
         list(APPEND ADIOS_LIBRARY_DIRS ${_LIBDIR})
     endforeach()
     # we could append ${CMAKE_PREFIX_PATH} now but that is not really necessary
+    list(APPEND ADIOS_LIBRARY_DIRS ${CMAKE_FIND_ROOT_PATH}/usr/lib)
 
     # determine whether found library links as serial only
     set(ADIOS_HAVE_SEQUENTIAL FALSE)
