From dfd65f23af7f907851d44ceb9e3aebdd05b4045f Mon Sep 17 00:00:00 2001
From: Axel Huebl <axel.huebl@plasma.ninja>
Date: Wed, 8 Jun 2022 00:56:41 -0700
Subject: [PATCH] 0.14.5 ADIOS1: Link Own `Task.cpp`

Backport of #1288 for #1287
---
 CMakeLists.txt | 6 ++++--
 1 file changed, 4 insertions(+), 2 deletions(-)

diff --git a/CMakeLists.txt b/CMakeLists.txt
index 141ec8f65..4282b3509 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -419,11 +419,13 @@ set(IO_SOURCE
 set(IO_ADIOS1_SEQUENTIAL_SOURCE
         src/auxiliary/Filesystem.cpp
         src/ChunkInfo.cpp
-        src/IO/ADIOS/ADIOS1IOHandler.cpp)
+        src/IO/ADIOS/ADIOS1IOHandler.cpp
+        src/IO/IOTask.cpp)
 set(IO_ADIOS1_SOURCE
         src/auxiliary/Filesystem.cpp
         src/ChunkInfo.cpp
-        src/IO/ADIOS/ParallelADIOS1IOHandler.cpp)
+        src/IO/ADIOS/ParallelADIOS1IOHandler.cpp
+        src/IO/IOTask.cpp)
 
 # library
 if(openPMD_BUILD_SHARED_LIBS)
