From ab29b2abf057fdfc303c7f5fbc04ea9414e528f3 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Franz=20P=C3=B6schel?= <franz.poeschel@gmail.com>
Date: Tue, 17 Feb 2026 11:23:23 +0100
Subject: [PATCH 1/2] Bump Pybind11 to 3.0.2

---
 cmake/dependencies/pybind11.cmake | 10 +++++-----
 1 file changed, 5 insertions(+), 5 deletions(-)

diff --git a/cmake/dependencies/pybind11.cmake b/cmake/dependencies/pybind11.cmake
index 68c0986e4f..643421df56 100644
--- a/cmake/dependencies/pybind11.cmake
+++ b/cmake/dependencies/pybind11.cmake
@@ -51,9 +51,9 @@ function(find_pybind11)
         mark_as_advanced(FETCHCONTENT_UPDATES_DISCONNECTED_FETCHEDpybind11)
     elseif(NOT openPMD_USE_INTERNAL_PYBIND11)
         if(openPMD_USE_PYTHON STREQUAL AUTO)
-            find_package(pybind11 2.13.0 CONFIG)
+            find_package(pybind11 3.0.0 CONFIG)
         elseif(openPMD_USE_PYTHON)
-            find_package(pybind11 2.13.0 CONFIG REQUIRED)
+            find_package(pybind11 3.0.0 CONFIG REQUIRED)
         endif()
         if(TARGET pybind11::module)
             message(STATUS "pybind11: Found version '${pybind11_VERSION}'")
@@ -67,10 +67,10 @@ set(openPMD_pybind11_src ""
     "Local path to pybind11 source directory (preferred if set)")
 
 # tarball fetcher
-set(openPMD_pybind11_tar "https://github.com/pybind/pybind11/archive/refs/tags/v2.13.6.tar.gz"
+set(openPMD_pybind11_tar "https://github.com/pybind/pybind11/archive/refs/tags/v3.0.2.tar.gz"
         CACHE STRING
         "Remote tarball link to pull and build pybind11 from if(openPMD_USE_INTERNAL_PYBIND11)")
-set(openPMD_pybind11_tar_hash "SHA256=e08cb87f4773da97fa7b5f035de8763abc656d87d5773e62f6da0587d1f0ec20"
+set(openPMD_pybind11_tar_hash "SHA256=2f20a0af0b921815e0e169ea7fec63909869323581b89d7de1553468553f6a2d"
         CACHE STRING
         "Hash checksum of the tarball of pybind11 if(openPMD_USE_INTERNAL_PYBIND11)")
 
@@ -78,7 +78,7 @@ set(openPMD_pybind11_tar_hash "SHA256=e08cb87f4773da97fa7b5f035de8763abc656d87d5
 set(openPMD_pybind11_repo "https://github.com/pybind/pybind11.git"
     CACHE STRING
     "Repository URI to pull and build pybind11 from if(openPMD_USE_INTERNAL_PYBIND11)")
-set(openPMD_pybind11_branch "v2.13.6"
+set(openPMD_pybind11_branch "v3.0.2"
     CACHE STRING
     "Repository branch for openPMD_pybind11_repo if(openPMD_USE_INTERNAL_PYBIND11)")
 

From 2603556d3d0c254169bada4173ece5dedf057166 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Franz=20P=C3=B6schel?= <franz.poeschel@gmail.com>
Date: Tue, 17 Feb 2026 11:35:55 +0100
Subject: [PATCH 2/2] Use cpp_function for defining properties

---
 src/binding/python/Iteration.cpp       | 24 ++++++++++++++----------
 src/binding/python/ParticleSpecies.cpp | 11 ++++++-----
 src/binding/python/Series.cpp          | 11 ++++++-----
 3 files changed, 26 insertions(+), 20 deletions(-)

diff --git a/src/binding/python/Iteration.cpp b/src/binding/python/Iteration.cpp
index 846861d262..8346aaaf63 100644
--- a/src/binding/python/Iteration.cpp
+++ b/src/binding/python/Iteration.cpp
@@ -94,18 +94,22 @@ void init_Iteration(py::module &m)
         .def("set_dt", &Iteration::setDt<double>)
         .def("set_time_unit_SI", &Iteration::setTimeUnitSI)
 
-        .def_readwrite(
+        .def_property_readonly(
             "meshes",
-            &Iteration::meshes,
-            py::return_value_policy::copy,
-            // garbage collection: return value must be freed before Iteration
-            py::keep_alive<1, 0>())
-        .def_readwrite(
+            py::cpp_function(
+                [](Iteration &i) { return i.meshes; },
+                py::return_value_policy::copy,
+                // garbage collection: return value must be freed before
+                // Iteration
+                py::keep_alive<1, 0>()))
+        .def_property_readonly(
             "particles",
-            &Iteration::particles,
-            py::return_value_policy::copy,
-            // garbage collection: return value must be freed before Iteration
-            py::keep_alive<1, 0>());
+            py::cpp_function(
+                [](Iteration &i) { return i.particles; },
+                py::return_value_policy::copy,
+                // garbage collection: return value must be freed before
+                // Iteration
+                py::keep_alive<1, 0>()));
 
     add_pickle(
         cl, [](openPMD::Series series, std::vector<std::string> const &group) {
diff --git a/src/binding/python/ParticleSpecies.cpp b/src/binding/python/ParticleSpecies.cpp
index 1af708d206..d619554220 100644
--- a/src/binding/python/ParticleSpecies.cpp
+++ b/src/binding/python/ParticleSpecies.cpp
@@ -48,12 +48,13 @@ void init_ParticleSpecies(py::module &m)
               return stream.str();
           })
 
-        .def_readwrite(
+        .def_property_readonly(
             "particle_patches",
-            &ParticleSpecies::particlePatches,
-            py::return_value_policy::copy,
-            // garbage collection: return value must be freed before Series
-            py::keep_alive<1, 0>());
+            py::cpp_function(
+                [](ParticleSpecies &ps) { return ps.particlePatches; },
+                py::return_value_policy::copy,
+                // garbage collection: return value must be freed before Series
+                py::keep_alive<1, 0>()));
     add_pickle(
         cl, [](openPMD::Series series, std::vector<std::string> const &group) {
             uint64_t const n_it = std::stoull(group.at(1));
diff --git a/src/binding/python/Series.cpp b/src/binding/python/Series.cpp
index 384462403d..ec2b38c867 100644
--- a/src/binding/python/Series.cpp
+++ b/src/binding/python/Series.cpp
@@ -497,12 +497,13 @@ this method.
         .def("set_iteration_format", &Series::setIterationFormat)
         .def("set_name", &Series::setName)
 
-        .def_readwrite(
+        .def_property_readonly(
             "iterations",
-            &Series::iterations,
-            py::return_value_policy::copy,
-            // garbage collection: return value must be freed before Series
-            py::keep_alive<1, 0>())
+            py::cpp_function(
+                [](Series &s) { return s.iterations; },
+                py::return_value_policy::copy,
+                // garbage collection: return value must be freed before Series
+                py::keep_alive<1, 0>()))
         .def(
             "read_iterations",
             [](Series &s) {
